### PR TITLE
no grad for vgg and vgg face to save gpu memory

### PR DESCRIPTION
**Pull request by eric8607242**
Vgg and vgg face with input x_hat need gradient to backprop to generator to update weight.
But vgg and vgg face with input x do not need gradient because the output is just like the groudtruth.

### CODE REVIEW

File: dataset/dataset.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 import cv2
 import matplotlib.pyplot as plt
 import numpy as np
-from face_alignment import FaceAlignment, LandmarksType
+#from face_alignment import FaceAlignment, LandmarksType
```
Comment (castelo-software) [inline dataset/dataset.py:16]: Why are you removing the import?

File: network/components.py [hunk 2]
```diff
@@ -66,6 +66,8 @@ def forward(self, x, mean_style, std_style):
         feature = x.view(B, C, -1)
 
         std_feat = (torch.std(feature, dim=2) + self.eps).view(B, C, 1)
+        #std_feat = torch.var(feature, dim=2) + self.eps
```
Comment (castelo-software) [inline network/components.py:69]: Not that important, but these comments are not needed.

File: network/components.py [hunk 3]
```diff
@@ -225,10 +227,21 @@ def forward(self, x, mean1, std1, mean2, std2):
...
         out = F.relu(out)
+        if torch.isnan(out[0, 0, 0, 0]):
+            print("-------nan  adarelu")
+        temp = out
         out = self.conv2(out)
+        if torch.isnan(out[0, 0, 0, 0]):
+            print("-------nan  ada3")
         out = self.in2(out, mean1, std1)
+        if torch.isnan(out[0, 0, 0, 0]):
+            print("-------nan  ada4")
```
Comment (castelo-software) [inline network/components.py]: These prints should be debug logs.

File: network/loss.py [hunk 4]
```diff
@@ -28,33 +27,37 @@ def loss_cnt(self, x, x_hat):
...
         vgg19_loss = 0
         for i in range(0, len(vgg19_x)):
             vgg19_loss += F.l1_loss(vgg19_x_hat[i], vgg19_x[i])
 
         # VGG Face Loss
         vgg_face_x_hat = self.VGG_FACE_AC(x_hat)
-
-        with torch.no_grad():
-            vgg_face_x = self.VGG_FACE_AC(x)
+        vgg_face_x = self.VGG_FACE_AC(x)
```
Comment (castelo-software) [inline network/loss.py]: I assume that you found out that you actually need gradients in these networks?

File: network/loss.py [hunk 5]
```diff
@@ -28,33 +27,37 @@ def loss_cnt(self, x, x_hat):
...
-    def loss_adv(self, r_x_hat):
-        return -r_x_hat.mean()
+    def loss_adv(self, r_x_hat, d_res, d_res_hat):
+        loss_fm = 0
+        for res, res_hat in zip(d_res, d_res_hat):
+            res = res.cuda(self.gpu)
+            res_hat = res_hat.cuda(self.gpu)
+            loss_fm += F.l1_loss(res, res_hat)
+
+        return -r_x_hat.mean() + loss_fm * config.LOSS_FM_WEIGHT
```
Comment (castelo-software) [inline network/loss.py:55]: It was concluded in a previous issue that Feature Matching is redundant and unnecessary. Why are you adding it back?

File: network/network.py [hunk 6]
```diff
@@ -167,13 +167,25 @@ def forward(self, y, e):
...
         out = self.res3(out, *self.slice_psi(psi_hat, 'res3'))
+        if torch.isnan(out[0, 0, 0, 0]):
+            print("nan------------Res2")
         out = self.res4(out, *self.slice_psi(psi_hat, 'res4'))
+        if torch.isnan(out[0, 0, 0, 0]):
+            print("nan------------Res3")
         out = self.res5(out, *self.slice_psi(psi_hat, 'res5'))
 
+        if torch.isnan(out[0, 0, 0, 0]):
+            print("nan------------Res4")
```
Comment (castelo-software) [inline network/network.py]: Same thing: debug logs, not prints

File: run.py [hunk 7]
```diff
@@ -18,11 +18,11 @@
...
-    'Embedder': 1,
-    'Generator': 0,
-    'Discriminator': 0,
-    'LossEG': 1,
-    'LossD': 1,
+    'Embedder': [0, 1],
+    'Generator': 2,
+    'Discriminator': 3,
+    'LossEG': 3,
+    'LossD': 3,
```
Comment (castelo-software) [inline run.py:25]: You don't need to commit this arrangement. Everyone's setup is different.

**Review by castelo-software — COMMENTED** [review]
@eric8607242,
Thanks a lot for your effort working on the model. Did you manage to fix the NaN problem at all? If you do, I'll look again at your branch, but for now I won't be merging this one.
